### DISCUSSION TITLE
router parameter in Application constructor

### DISCUSSION THREAD

**Discussion by netwang** [Neoteroi/BlackSheep#464] [Q&A] [answered]
Application(
     services=services,
     show_error_details=configuration.show_error_details,
     router=RouterWithPrefix(f'/{configuration.App_Name}/') if configuration.production else None,
)

class RouterWithPrefix(Router):
      """Router With Prefix."""

       def __init__(self, prefix: str) -> None:
            super().__init__()

            self.__prefix = prefix

      def add(self, method: str, pattern: AnyStr, handler: Any) -> None:
            """Add."""
            if isinstance(pattern, str):
                  pattern = (self.__prefix + pattern).replace('//', '/')
           else:
                 pattern = (self.__prefix.encode() + pattern).replace(b'//', b'/')

          super().add(method, pattern, handler)

Above code is for BlackSheep V1.2.18 to add prefix to all routes,. It changes
          http://localhost:1400/login
to
          http://localhost:1400/QMOSDS/login

I updated to BlackSheep 2.0.4. Above cold will not work when configuration.production is True.

Above cold will work when configuration.production is False. Use default_router

    if router is None:
        router = default_router if env_settings.use_default_router else Router()

How to add prefix to all routes in BlackSheep 2.0.4?? Thanks!

**Comment by RobertoPrevato** [comment, marked as answer]
Hi @netwang 

The below works fine:

```python
import os
from blacksheep import Application, Router


class RouterWithPrefix(Router):
    """Router With Prefix."""

    def __init__(self, prefix: str) -> None:
        super().__init__()

        self.__prefix = prefix

    def add(self, method: str, pattern, handler) -> None:
        """Add."""
        if isinstance(pattern, str):
            pattern = (self.__prefix + pattern).replace("//", "/")
        else:
            pattern = (self.__prefix.encode() + pattern).replace(b"//", b"/")

        super().add(method, pattern, handler)


prod = os.environ.get("PROD") == "1"


app = Application(router=RouterWithPrefix(f"/example/") if prod else None)


@app.router.get("/")
def home():
    return "Home!"
```

If you define your own router, you **cannot** use the methods imported from BlackSheep, because they are the methods of the default router. You need to use `get`, `post`, etc. of the router you are using.

For example, create a module in your application in `app/router.py`:

```python
# app/router.py
import os

if os.environ.get("PROD") in {"1", "true"}:  # if configuration.production: ...
    router = RouterWithPrefix(...)
else:
    router = Router()


head = router.head
get = router.get
post = router.post
put = router.put
patch = router.patch
delete = router.delete
trace = router.trace
options = router.options
connect = router.connect
ws = router.ws
route = router.route 
```

Then use the `get`, `post`, etc. methods imported from your module to define routes, not the default methods.

**Reply by netwang** [reply]
Problem is that I use below function to configure application. RouterWithPrefix should use the configured app_name. 

def configure_application(
    services: Container,
    context: ServicesRegistrationContext,
    configuration: Configuration,
) -> Application:
    """Configure app."""
    # if not sys.warnoptions:
    #     warnings.simplefilter('error')

    app_name = configuration.App_Name

    app = Application(
        services=services,
        show_error_details=configuration.show_error_details,
        router=RouterWithPrefix(f'/{app_name}/') if configuration.production else None,
    )


I can not use below solution:

if os.environ.get("PROD") in {"1", "true"}:  # if configuration.production: ...
    router = RouterWithPrefix(...)
else:
    router = Router()


head = router.head
get = router.get

**Reply by RobertoPrevato** [reply]
My example is to describe how to use a custom router, and my example is working.
Load configuration in the way you prefer, and configure `app_name` in your RouterWithPrefix.

**Comment by netwang** [comment]
all My projects were split into 2 parts:

- application part
- common library (libquad)

both application part and common library (libquad) have following code

from blacksheep.server.controllers import post

class xxxController(ControllerBase):
     xxxxxxxx

Common controllers are put into the common library (libquad). That's why I can not use below solution:

if os.environ.get("PROD") in {"1", "true"}: # if configuration.production: ...
router = RouterWithPrefix(...)
else:
router = Router()

head = router.head
get = router.get


I can not use app.router either.

@app.router.get("/")
def home():
    return "Home!"

**Reply by netwang** [reply]
I changed route.pattern in after_start handler

    app = Application(
        services=services,
        show_error_details=configuration.show_error_details,
    )

app.after_start += __print_routes

async def __print_routes(application: Application) -> None:
   """Print routes."""
   global app_name

   for method, routes in application.router.routes.items():
       print(method.decode('utf-8'))
       for route in routes:
           route.pattern = f'/{app_name}{route.pattern.decode("utf-8")}'.encode('utf-8')
           print('    ', route.pattern.decode('utf-8'))

But the following don't change. How to change them to add the route prefix?
route.full_pattern
route.rx
route.value_patterns

**Comment by netwang** [comment]
fixed

async def __print_routes(application: Application) -> None:
    """Print routes."""
    global app_name
    global production

    if production:
        routesDict: dict[bytes, list[Route]] = defaultdict(list)

        for method, routes in application.router.routes.items():
            print(method.decode('utf-8'))
            routesNew: list[Route] = []

            for route in routes:
                pattern = route.pattern.decode('utf-8')
                if pattern.startswith('/ws/') or pattern.startswith('/api/'):
                    newRoute = Route(f'/{app_name}{pattern}', route.handler)
                    routesNew.append(newRoute)
                    print('    ', newRoute.pattern.decode('utf-8'))
                else:
                    routesNew.append(route)
                    print('    ', route.pattern.decode('utf-8'))

            routesDict[method] = routesNew

        application.router.routes = routesDict
    else:
        for method, routes in application.router.routes.items():
            print(method.decode('utf-8'))
            for route in routes:
                print('    ', route.pattern.decode('utf-8'))
